### PR TITLE
native: allow for native to be resetable via SIGUSR1

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Based on the discussion sparked by this comment https://github.com/RIOT-OS/RIOT/issues/12448#issuecomment-541732537 I decided to provide a simple PoC of how to make `native` resetable.

### Testing procedure
Start a `native` instance in one terminal (application is arbitrary), find out its PID in another terminal, and then use the PID to reset e.g.:

```
$ pgrep -f native/hello-world
96937
$ DEBUG_ADAPTER_ID=96937 make -C examples/hello-world reset
```
The `native` instance should then reboot (noticeable by a big fat `!! REBOOT !!` printed)

### Issues/PRs references
https://github.com/RIOT-OS/RIOT/issues/12448#issuecomment-541732537

### CODE REVIEW

**Comment by fjmolinas** [comment]
Test, works as advertised.

```
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.01-devel-194-g943fd-pr-12517)
Hello World!
You are running RIOT on a(n) native board.
This board features a(n) native MCU.


                !! REBOOT !!

RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.01-devel-194-g943fd-pr-12517)
Hello World!
You are running RIOT on a(n) native board.
This board features a(n) native MCU.

```

**Comment by smlng** [comment]
on FreeBSD I get:

```
~/devel/RIOT/examples/hello-world/bin/native/hello-world.elf: native_irq_handler: ignoring SIGUSR1
```

**Comment by smlng** [comment]
comes from here https://github.com/RIOT-OS/RIOT/blob/master/cpu/native/irq_cpu.c#L266

File: cpu/native/startup.c [hunk 1]
```diff
@@ -577,6 +586,10 @@ __attribute__((constructor)) static void startup(int argc, char **argv, char **e
     periph_init();
     board_init();
 
+    /* should not error as SIGUSR is a valid error code and EINVAL is the only
+     * specified error code => don't check return value */
+    signal(SIGUSR1, reset_handler);
```
Comment (smlng) [inline cpu/native/startup.c]: I think you need to (should) use: https://github.com/RIOT-OS/RIOT/blob/master/cpu/native/irq_cpu.c#L412 instead

Comment (smlng) [reply cpu/native/startup.c]: or additionally?

Comment (miri64) [reply cpu/native/startup.c]: Might be.

File: cpu/native/startup.c [hunk 2]
```diff
@@ -390,6 +392,13 @@ extern init_func_t __init_array_start;
 extern init_func_t __init_array_end;
 #endif
 
+static void reset_handler(int signo)
```
Comment (smlng) [inline cpu/native/startup.c]: I changed this to 
```
static void _reset_handler(void)
{
    pm_reboot();
}
```

and added:

```
register_interrupt(SIGUSR1, _reset_handler);
```

below and it works on FreeBSD, the `signal(...)` is not needed

**Comment by smlng** [comment]
with the proposed changes it looks good:

```
/devel/RIOT/examples/hello-world/bin/native/hello-world.elf
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.01-devel-194-g943fd-HEAD)
Hello World!
You are running RIOT on a(n) native board.
This board features a(n) native MCU.


                !! REBOOT !!

RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.01-devel-194-g943fd-HEAD)
Hello World!
You are running RIOT on a(n) native board.
This board features a(n) native MCU.
```

**Comment by miri64** [comment]
I will apply the changes @smlng proposed and retest on Linux.

**Comment by miri64** [comment]
> I will apply the changes @smlng proposed and retest on Linux.

Done and still works on Linux. Thanks for testing on FreeBSD, @smlng!

File: cpu/native/include/native_internal.h [hunk 3]
```diff
@@ -64,6 +64,13 @@ typedef void (*_native_callback_t)(void);
  * @cond INTERNAL
  * internal functions
  */
+/**
+ * @brief   Don't use Linux' `sighandler_t` as it is not POSIX compliant
+ *
+ * @see e.g. https://www.freebsd.org/cgi/man.cgi?query=signal&section=3
+ */
+typedef void (*real_sig_t) (int);
```
Comment (smlng) [inline cpu/native/include/native_internal.h]: not needed anymore, right?

File: cpu/native/include/native_internal.h [hunk 4]
```diff
@@ -121,6 +128,7 @@ extern int (*real_setitimer)(int which, const struct itimerval
         *__restrict value, struct itimerval *__restrict ovalue);
 extern int (*real_setsid)(void);
 extern int (*real_setsockopt)(int socket, ...);
+extern real_sig_t (*real_signal)(int sig, real_sig_t handler);
```
Comment (smlng) [inline cpu/native/include/native_internal.h]: same here?

File: cpu/native/include/native_internal.h (hunk 3 above)
Comment (miri64) [reply cpu/native/include/native_internal.h]: Oh yeah, sorry.... still on my first coffee ^^"

File: cpu/native/syscalls.c [hunk 5]
```diff
@@ -87,6 +87,7 @@ int (*real_setitimer)(int which, const struct itimerval
         *restrict value, struct itimerval *restrict ovalue);
 int (*real_setsid)(void);
 int (*real_setsockopt)(int socket, ...);
+real_sig_t (*real_signal)(int sig, real_sig_t handler);
```
Comment (smlng) [inline cpu/native/syscalls.c]: also not needed anymore?

**Comment by smlng** [comment]
when done, please squash and trigger Murdock please

File: cpu/native/include/native_internal.h (hunk 3 above)
Comment (smlng) [reply cpu/native/include/native_internal.h]: enjoy that first ☕️

**Comment by miri64** [comment]
Cleaned out now unnecessary changes and squashed.

**Comment by kaspar030** [comment]
> tested ACK: Linux and FreeBSD

I guess FreeBSD implies OSX?

**Comment by smlng** [comment]
RIOT does not work on latest macOS, bc the latter is 64bit only now - so no macOS right now

**Comment by cgundogan** [comment]
>RIOT does not work on latest macOS, bc the latter is 64bit only now - so no macOS right now

@x3ro are you still working on a 64-bit native port? (:

**Comment by miri64** [comment]
> > tested ACK: Linux and FreeBSD
> 
> I guess FreeBSD implies OSX?

There is I think twenty years of divergence between FreeBSD and Darwin ^^, but okay :-)

**Pull request merged**